background/index.ts

```diff
@@ -94,6 +97,172 @@ async function main() {
     { pendingTransaction: {} }
   )
 
+  const indexBalanceOnBlock = async () => {
+    const blockSubscriberContext: {
+      provider?: JsonRpcProvider
+      blockSubscriber?: Listener
+    } = {}
+
+    // Syncs account balances using multicall
+    const updateAccountBalances = async (
+      accounts: Account[],
+      provider: JsonRpcProvider
+    ) => {
+      // Use multicall provider to optimize RPC calls
+      const multicallProvider = MulticallWrapper.wrap(provider)
+
+      // Store balance update promises
+      const tokenQueries = []
+
+      // Update balances for each account
+      accounts.forEach((account) => {
+        const {
+          id,
+          tokens,
+          chainId,
+          balance: accountBalance,
+          address: accountAddress
+        } = account
+
+        tokenQueries.push(
+          (async () => {
+            const nativeBalance =
+              await multicallProvider.getBalance(accountAddress)
+
+            // Update balance if changed
+            if (number.toBigInt(accountBalance) !== nativeBalance) {
+              storage.set((state) => {
+                state.account[id].balance = number.toHex(nativeBalance)
```

Comment:
I am thinking about that can we set all these balances (including tokens') at once to avoid triggering storage update and subscribers too frequently?

The process would be (1) query all native & token balances (2) on balances returning, call `storage.set` only once to update them all

Reply:
Thank you! Modified in commit [https://github.com/pilagod/waallet-extension/pull/202/commits/47ec7f31bb806633450f94384cf2d98d3a78333f](https://github.com/pilagod/waallet-extension/pull/202/commits/47ec7f31bb806633450f94384cf2d98d3a78333f).